pkg/controller/orchestrator/orchestrator_reconcile.go

```diff
@@ -0,0 +1,478 @@
+/*
+Copyright 2018 Pressinfra SRL
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package orchestrator
+
+import (
+	"fmt"
+	"time"
+
+	core "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	api "github.com/presslabs/mysql-operator/pkg/apis/mysql/v1alpha1"
+	wrapcluster "github.com/presslabs/mysql-operator/pkg/controller/internal/mysqlcluster"
+	orc "github.com/presslabs/mysql-operator/pkg/orchestrator"
+)
+
+const (
+	healtyMoreThanMinutes        = 10
+	defaultMaxSlaveLatency int64 = 30
+	mysqlPort                    = 3306
+)
+
+type orcUpdater struct {
+	cluster   *wrapcluster.MysqlCluster
+	recorder  record.EventRecorder
+	orcClient orc.Interface
+}
+
+// Syncer interface defines Sync method
+type Syncer interface {
+	Sync() error
+}
+
+// NewOrcUpdater returns a syncer that updates cluster status from orchestrator.
+func NewOrcUpdater(cluster *api.MysqlCluster, r record.EventRecorder, orcClient orc.Interface) Syncer {
+	return &orcUpdater{
+		cluster:   wrapcluster.NewMysqlClusterWrapper(cluster),
+		recorder:  r,
+		orcClient: orcClient,
+	}
+}
+
+func (ou *orcUpdater) Sync() error {
+	// get instances from orchestrator
+	var (
+		instances  InstancesSet
+		err        error
+		recoveries []orc.TopologyRecovery
+	)
+
+	if instances, err = ou.orcClient.Cluster(ou.cluster.GetClusterAlias()); err != nil {
+		log.Error(err, "can't get instances from orchestrator", "alias", ou.cluster.GetClusterAlias(), "cluster", ou.cluster)
+	}
+
+	if len(instances) == 0 {
+		log.Info("no instances in orchestrator", "clusterAlias", ou.cluster.GetClusterAlias(), "cluster", ou.cluster)
+	}
+
+	// register nodes in orchestrator if needed, or remove nodes from status
+	instances = ou.registerUnregisterNodesInOrc(instances)
+
+	// remove nodes that are not in orchestrator
+	ou.removeNodeConditionNotInOrc(instances)
+
+	// set readonly in orchestrator if needed
+	if err = ou.markReadOnlyNodesInOrc(instances); err != nil {
+		log.Error(err, "error setting Master readOnly/writable", "instances", instances, "cluster", ou.cluster)
```

Comment:
logging the entire cluster is a bit too verbose. We should log only name/namespace key.